File: ios/BACtrackReactNativeDemo/BactrackModule.m
//
//  BatteryTest.m
//  BACtrackReactNativeDemo
//
//  Created by BACtrack
//

#import <Foundation/Foundation.h>
#import "BactrackModule.h"
#import "BACTrack.h"
#import <UIKit/UIKit.h>


@interface BactrackModule () <BacTrackAPIDelegate>
{
  BacTrackAPI * mBacTrack;
}
@end

@implementation BactrackModule
{
  bool hasListeners;
}
- (dispatch_queue_t)methodQueue
{
  return dispatch_get_main_queue();
}

+ (BOOL) requiresMainQueueSetup
{
  return YES;
}
// Will be called when this module's first listener is added.
-(void)startObserving {
  hasListeners = YES;
  // Set up any upstream listeners or background tasks as necessary
}

// Will be called when this module's last listener is removed, or on dealloc.
-(void)stopObserving {
  hasListeners = NO;
  // Remove upstream listeners, stop unnecessary background tasks
}

RCT_EXPORT_MODULE();


- (NSArray<NSString *> *)supportedEvents
{
  return @[@"onKeyDeclined", @"onKeyAuthorized", @"onSerialDetected", @"onUuid",  @"onUseCount", @"onBacTrackCountdown", @"onBacResult", @"onDeviceConnected", @"onBacDisconnected",@"OnBatteryVoltage", @"onBatteryLevel", @"onBacTrackStart", @"onBacTrackBlow", @"onBacTrackAnalyzing", @"onError" ];
}


- (id) init
{
  return [super init];
}

-(void)BacTrackAPIKeyAuthorized
{
  NSLog(@"BACtrack: API Key Authorized");
  if(hasListeners) {
    [self sendEventWithName:@"onKeyAuthorized" body: @"true"];
  }
}

//API Key declined for some reason
-(void)BacTrackAPIKeyDeclined:(NSString *)errorMessage
{
  NSLog(@"BACtrack: API Key Declined");
  if(hasListeners) {
    [self sendEventWithName:@"onKeyDeclined" body: errorMessage];
  }
}


-(void)BacTrackSerial:(NSString *)serial_hex
{
  NSLog(@"BACtrack: Serial Number %@", serial_hex);
  if(hasListeners) {
    [self sendEventWithName:@"onSerialDetected" body: serial_hex];
  }
}

-(void)BacTrackUseCount:(NSNumber*)number
{
  NSLog(@"BACtrack: Device Use Count: %d", number.intValue);
  if(hasListeners) {
    [self sendEventWithName:@"onUseCount" body: number.stringValue];
  }
}


-(void)BacTrackError:(NSError*)error
{
  NSLog(@"BACtrack: Error %@", error.localizedDescription);
  if(hasListeners) {
    [self sendEventWithName:@"onError" body: @(error.code).stringValue];
  }
}

-(void)BacTrackCountdown:(NSNumber *)seconds executionFailure:(BOOL)failure
{
  NSLog(@"BACtrack: Countdown: %d", seconds.intValue);

  if(hasListeners) {
    [self sendEventWithName:@"onBacTrackCountdown" body: seconds.stringValue];
  }
}


-(void)BacTrackResults:(CGFloat)bac
{
  NSLog(@"BACtrack: BacTrackResults result %f", bac);
  if(hasListeners) {
    [self sendEventWithName:@"onBacResult" body: [NSString stringWithFormat:@"%f", bac]];
  }
}

-(void)BacTrackConnected:(BACtrackDeviceType)device
{
  if (device == BACtrackDeviceType_Mobile ) {
    NSLog(@"BACtrack: Connected to device: bactrack-mobile");
  } else if (device == BACtrackDeviceType_MobileV2 ) {
    NSLog(@"BACtrack: Connected to device: bactrack-mobile-v2");
  } else if(device == BACtrackDeviceType_C6){
    NSLog(@"BACtrack: Connected to device: bactrack-c6");
  } else if(device == BACtrackDeviceType_C8){
    NSLog(@"BACtrack: Connected to device: bactrack-c8");
  } else {
    NSLog(@"BACtrack: Connected to device: unknown :%d", (int)device);
  }

  if(hasListeners) {
    [self sendEventWithName:@"onDeviceConnected" body: [NSString stringWithFormat:@"%d", ((int)device)]];
  }
}

-(void)BacTrackDisconnected
{
  NSLog(@"BACtrack: Disconnected");
  if(hasListeners) {
    [self sendEventWithName:@"onBacDisconnected" body:@"true"];
  }
}

-(void)BacTrackConnectTimeout
{
  NSLog(@"BACtrack: Connection Timeout");
}

-(NSTimeInterval)BacTrackGetTimeout
{
  return 10;
}


-(void)BacTrackFoundBreathalyzer:(Breathalyzer*)breathalyzer
{
  NSLog(@"BACtrack: FoundBreathalyzer: %@", breathalyzer.peripheral.name.description);
  if(hasListeners) {
    [self sendEventWithName:@"onUuid" body:breathalyzer.uuid];
  }
}

- (void) BacTrackBatteryVoltage:(NSNumber *)number
{
  NSLog(@"BACtrack: Battery Voltage: %@", number.stringValue);

  if(hasListeners) {
    [self sendEventWithName:@"OnBatteryVoltage" body: number.stringValue];
  }
}

- (void) BacTrackBatteryLevel:(NSNumber *)number
{
  NSLog(@"BACtrack: Battery Level: %@",  number.stringValue);

  if(hasListeners) {
    [self sendEventWithName:@"onBatteryLevel" body:number.stringValue];
  }
}


- (void) BacTrackStart {
  NSLog(@"BACtrack: Start");
  if(hasListeners) {
    [self sendEventWithName:@"onBacTrackStart" body:@"true"];
  }
}


- (void)BacTrackBlow:(NSNumber*)breathFractionRemaining {
  NSLog(@"BACtrack: BacTrackBlow breath fraction remaining %@", breathFractionRemaining.stringValue);
  if(hasListeners) {
    [self sendEventWithName:@"onBacTrackBlow" body: breathFractionRemaining.stringValue];
  }
}


- (void) BacTrackAnalyzing {
    NSLog(@"BACtrack: Analyzing results");
    if(hasListeners) {
      [self sendEventWithName:@"onBacTrackAnalyzing" body:@"true"];
    }
}

-(void)dealloc
{
  if (mStopScanTimer)
  {
    [mStopScanTimer invalidate];
    mStopScanTimer = nil;
  }
}

// ACTION METHODS

RCT_EXPORT_METHOD(initBACtrack) {
    NSLog(@"BACtrack: Init BACtrack");
    mBacTrack = [[BacTrackAPI alloc] initWithDelegate:self AndAPIKey:@"You API Key goes here"];
    [mBacTrack startScan];

}

RCT_EXPORT_METHOD(connectToNearestBreathalyzer) {
    NSLog(@"BACtrack: Connect To Nearest Breathalyzer");
    [mBacTrack connectToNearestBreathalyzer];

}


RCT_EXPORT_METHOD(getBatteryLevel){
  NSLog(@"BACtrack: Getting battery level and voltage");
  [mBacTrack getBreathalyzerBatteryLevel];
}

RCT_EXPORT_METHOD(getSerialNumber){
  NSLog(@"BACtrack: Getting serial number");
  [mBacTrack getBreathalyzerSerialNumber];
}


RCT_EXPORT_METHOD(startTestCountdown){
  NSLog(@"BACtrack: Starting Test Countdown");
  [mBacTrack startCountdown];
}


RCT_EXPORT_METHOD(disconnect) {
  NSLog(@"BACtrack: Disconnecting");
  [mBacTrack disconnect];
}

@end

File: android/app/src/main/java/com/bactrackreactnativedemo/BactrackModule.java
package com.bactrackreactnativedemo;

import android.util.Log;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import BreathalyzerSDK.API.BACtrackAPI;
import BreathalyzerSDK.API.BACtrackAPICallbacks;
import BreathalyzerSDK.Constants.BACTrackDeviceType;
import BreathalyzerSDK.Constants.BACtrackUnit;
import BreathalyzerSDK.Exceptions.BluetoothLENotSupportedException;
import BreathalyzerSDK.Exceptions.BluetoothNotEnabledException;
import BreathalyzerSDK.Exceptions.LocationServicesNotEnabledException;

public class BactrackModule extends ReactContextBaseJavaModule {
    private static ReactApplicationContext reactContext;
    private static final String BACTRACK_API_KEY = "Your API Key goes here";
    private BACtrackAPI mAPI;
    BactrackModule(ReactApplicationContext context) {
        super(context);
        reactContext = context;
    }

    @Override
    public String getName() {
        return "BactrackModule";
    }

    private void sendEvent(
        ReactContext reactContext,
        String eventName,
        String params) {
        reactContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
            .emit(eventName, params);
    }




    private final BACtrackAPICallbacks mCallbacks = new BACtrackAPICallbacks() {
        @Override
        public void BACtrackAPIKeyDeclined(String errorMessage) {
            sendEvent(reactContext, "onKeyDeclined", errorMessage);
        }

        @Override
        public void BACtrackAPIKeyAuthorized() {
            sendEvent(reactContext, "onKeyAuthorized", "true");
        }

        @Override
        public void BACtrackConnected(BACTrackDeviceType bacTrackDeviceType) {
            sendEvent(reactContext, "onDeviceConnected", bacTrackDeviceType.getDisplayName());
        }

        @Override
        public void BACtrackDidConnect(String s) {
            Log.d("DEBUG_TAG", "BACtrackDidConnect: " + s);
        }

        @Override

        public void BACtrackDisconnected() {
            sendEvent(reactContext, "onBacDisconnected", "true");
        }

        @Override
        public void BACtrackConnectionTimeout() {
            Log.d("DEBUG_TAG", "BACtrackConnectionTimeout: ");
        }

        @Override
        public void BACtrackUnits(BACtrackUnit unit) {
            Log.d("DEBUG_TAG", "BACtrackUnits: " + unit.name());
        }

        @Override
        public void BACtrackFoundBreathalyzer(BACtrackAPI.BACtrackDevice BACtrackDevice) {
            Log.d("DEBUG_TAG", "BACtrackFoundBreathalyzer: " + BACtrackDevice.device);
        }

        @Override
        public void BACtrackCountdown(int currentCountdownCount) {
            sendEvent(reactContext, "onBacTrackCountdown", String.valueOf((currentCountdownCount)));
        }

        @Override
        public void BACtrackStart() {
            sendEvent(reactContext, "onBacTrackStart", "true");
        }

        @Override
        public void BACtrackBlow(float blowFractionRemaining) {
            sendEvent(reactContext, "onBacTrackBlow", String.valueOf(blowFractionRemaining));
        }

        @Override
        public void BACtrackAnalyzing() {
            sendEvent(reactContext, "onBacTrackAnalyzing", "true");
        }

        @Override
        public void BACtrackResults(float measuredBac) {
            sendEvent(reactContext, "onBacResult", String.valueOf(measuredBac));
        }

        @Override
        public void BACtrackFirmwareVersion(String version) {
            sendEvent(reactContext, "onFirmwareVersion", version);
        }

        @Override
        public void BACtrackSerial(String serialHex) {
            sendEvent(reactContext, "onSerialDetected", serialHex);
        }

        @Override
        public void BACtrackUseCount(int useCount) {
            sendEvent(reactContext, "onUseCount", String.valueOf(useCount));
        }

        @Override
        public void BACtrackBatteryVoltage(float voltage) {
            sendEvent(reactContext, "OnBatteryVoltage", String.valueOf(voltage));

        }

        @Override
        public void BACtrackBatteryLevel(int level) {
            sendEvent(reactContext, "onBatteryLevel", String.valueOf(level));
        }

        @Override
        public void BACtrackError(int errorCode) {
            sendEvent(reactContext, "onError", String.valueOf(errorCode));
        }
    };

    @ReactMethod
    public void initBACtrack() {
        try {
            mAPI = new BACtrackAPI(reactContext, mCallbacks, BACTRACK_API_KEY);
        } catch (BluetoothLENotSupportedException | BluetoothNotEnabledException |
                 LocationServicesNotEnabledException e) {
            sendEvent(reactContext, "onError", "Permissions not granted");
            e.printStackTrace();
        }
    }

    @ReactMethod
    public void connectToNearestBreathalyzer() {
        if (mAPI != null) {
            mAPI.connectToNearestBreathalyzer();
        }
    }

    @ReactMethod
    public void getBatteryLevel() {
        if (mAPI != null) {
            mAPI.getBreathalyzerBatteryVoltage();
        }
    }

    @ReactMethod
    public void getSerialNumber() {
        if (mAPI != null) {
            mAPI.getSerialNumber();
        }
    }

    @ReactMethod
    public void startTestCountdown() {
        if (mAPI != null) {
            mAPI.startCountdown();
        }
    }

    @ReactMethod
    public void disconnect() {
        if (mAPI != null) {
             mAPI.disconnect();
        }
    }


    @ReactMethod
    public void addListener(String eventName) {
        Log.d("DEBUG_TAG", "addListener: " + eventName);

    }

    @ReactMethod
    public void removeListeners(Integer count) {
        Log.d("DEBUG_TAG", "removeListeners: " + count);
    }

}



